cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java

```diff
@@ -100,717 +108,167 @@ public EntityTypeSimpleName apply(String input) {
-    StreamViewId view = new StreamViewId(namespaceId, streamId, viewId);
-    metadataAdmin.removeTags(view);
-    responder.sendString(HttpResponseStatus.OK,
-                         String.format("Tags for view %s deleted successfully.", view));
-  }
-
-  @DELETE
-  @Path("/namespaces/{namespace-id}/streams/{stream-id}/metadata/tags/{tag}")
-  public void removeStreamTag(HttpRequest request, HttpResponder responder,
-                              @PathParam("namespace-id") String namespaceId,
-                              @PathParam("stream-id") String streamId,
-                              @PathParam("tag") String tag) throws NotFoundException {
-    StreamId stream = new StreamId(namespaceId, streamId);
-    metadataAdmin.removeTags(stream, tag);
-    responder.sendString(HttpResponseStatus.OK,
-                         String.format("Tag %s for stream %s deleted successfully.", tag, stream));
+                             @QueryParam("query") String searchQuery,
+                             @QueryParam("target") List<String> targets,
+                             @QueryParam("sort") @DefaultValue("") String sort,
+                             @QueryParam("offset") @DefaultValue("0") int offset,
+                             // 2147483647 is Integer.MAX_VALUE
+                             @QueryParam("limit") @DefaultValue("2147483647") int limit,
+                             @QueryParam("numCursors") @DefaultValue("0") int numCursors,
+                             @QueryParam("cursor") @DefaultValue("") String cursor,
+                             @QueryParam("showHidden") @DefaultValue("false") boolean showHidden,
+                             @Nullable @QueryParam("entityScope") String entityScope) throws Exception {
+    if (searchQuery == null || searchQuery.isEmpty()) {
+      throw new BadRequestException("query is not specified");
+    }
+    Set<EntityTypeSimpleName> types = Collections.emptySet();
+    if (targets != null) {
+      types = ImmutableSet.copyOf(targets.stream().map(STRING_TO_TARGET_TYPE).collect(Collectors.toList()));
+    }
+    SortInfo sortInfo = SortInfo.of(URLDecoder.decode(sort, "UTF-8"));
+    if (SortInfo.DEFAULT.equals(sortInfo)) {
+      if (!(cursor.isEmpty()) || 0 != numCursors) {
+        throw new BadRequestException("Cursors are not supported when sort info is not specified.");
+      }
+    }
+    try {
+      MetadataSearchResponse response =
+        metadataAdmin.search(namespaceId, URLDecoder.decode(searchQuery, "UTF-8"), types,
+                             sortInfo, offset, limit, numCursors, cursor, showHidden, validateEntityScope(entityScope));
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(response, MetadataSearchResponse.class));
+    } catch (Exception e) {
+      // if MetadataDataset throws an exception, it gets wrapped
+      if (Throwables.getRootCause(e) instanceof BadRequestException) {
```

Comment:
Is this correct? This will get the root... meaning if you have

```
RuntimeException
  |- BadRequestException
    |- IllegalArgumentException
```

The `getRootCause` call will give you `IllegalArgumentException`.

Reply:
Please respond to this comment